pubsub/rocketmq/rocketmq.go

```diff
@@ -140,207 +265,199 @@ func (r *rocketMQ) Features() []pubsub.Feature {
-	// reset synchronization
-	if r.consumer != nil {
-		r.logger.Infof("re-initializing the consumer")
-		_ = r.consumer.Shutdown()
-		r.consumer = nil
-	} else {
-		r.logger.Infof("initializing the consumer")
+func buildMessageSelector(req pubsub.SubscribeRequest) (*mqc.MessageSelector, error) {
+	if req.Metadata == nil {
+		req.Metadata = make(map[string]string)
 	}
-
-	r.consumer, err = r.setUpConsumer()
-	if err != nil {
-		r.consumer = nil
-		return err
+	mqExpr := req.Metadata[metadataRocketmqExpression]
+	mqType := req.Metadata[metadataRocketmqType]
+
+	var ExpressionType mqc.ExpressionType
+	if mqType == "" || strings.EqualFold(mqType, string(mqc.TAG)) {
+		ExpressionType = mqc.TAG
+	} else if strings.EqualFold(mqType, string(mqc.SQL92)) {
+		ExpressionType = mqc.SQL92
+	} else {
+		return nil, fmt.Errorf("rocketmq msg type invalid: %s, expected value is 'tag' or 'sql92' or ''", mqType)
 	}
 
-	// Invoke onConnReady so changes to the topics can be made safely
-	onConnRready()
+	return &mqc.MessageSelector{
+		Type:       ExpressionType,
+		Expression: mqExpr,
+	}, nil
+}
 
-	for topic, data := range r.topics {
-		cb := r.adaptCallback(topic, r.metadata.ConsumerGroup, string(data.selector.Type), data.selector.Expression, data.handler)
-		err = r.consumer.Subscribe(topic, data.selector, cb)
-		if err != nil {
-			r.logger.Errorf("subscribe topic:%v failed,error:%v", topic, err)
-			continue
+func (r *rocketMQ) buildPubsubMessage(topic, mqType, mqExpr string, msg *primitive.MessageExt) (*pubsub.NewMessage, error) {
+	cloudEventsMap := pubsub.NewCloudEventsEnvelope(msg.MsgId, msg.StoreHost, "", "", msg.Topic, r.name, r.metadata.ContentType, msg.Body, "", "")
+	cloudEventsMap[primitive.PropertyKeys] = msg.GetKeys()
+	cloudEventsMap[primitive.PropertyShardingKey] = msg.GetShardingKey()
+	cloudEventsMap[primitive.PropertyTags] = msg.GetTags()
+	cloudEventsMap[primitive.PropertyMsgRegion] = msg.GetRegionID()
+	for k, v := range msg.GetProperties() {
+		if _, ok := r.msgProperties[k]; ok {
+			cloudEventsMap[k] = v
+		}
+		if strings.EqualFold(k, pubsub.TraceIDField) {
+			cloudEventsMap[pubsub.TraceIDField] = v
 		}
 	}
-
-	err = r.consumer.Start()
+	dataBytes, err := json.Marshal(cloudEventsMap)
 	if err != nil {
-		return fmt.Errorf("consumer start failed. %w", err)
+		return nil, err
+	}
+	metadata := map[string]string{
+		metadataRocketmqType:          mqType,
+		metadataRocketmqExpression:    mqExpr,
+		metadataRocketmqConsumerGroup: r.metadata.ProducerGroupName,
+	}
+	if msg.Queue != nil {
+		metadata[metadataRocketmqBrokerName] = msg.Queue.BrokerName
+		metadata[metadataRocketmqQueueID] = strconv.Itoa(msg.Queue.QueueId)
 	}
+	return &pubsub.NewMessage{
+		Topic:    topic,
+		Data:     dataBytes,
+		Metadata: metadata,
+	}, nil
+}
 
-	return nil
+func (r *rocketMQ) consumeMessageOrderly(topic string, selector *mqc.MessageSelector, handler pubsub.Handler) func(ctx context.Context, msgs ...*primitive.MessageExt) (mqc.ConsumeResult, error) {
```

Comment:
What is the difference betwen `consumeMessageOrderly` and `consumeMessageCurrently`? I can't see any difference in code?

Reply:
These two methods return different results. ConsumeMessageOrderly return result is SuspendCurrentQueueAMoment or ConsumeSuccess, cannot be ConsumeRetryLater